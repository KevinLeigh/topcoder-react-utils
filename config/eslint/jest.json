{
  "extends": "./default.json",
  "env": {
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "global-require": 0,
    "import/no-dynamic-require": 0
  }
}