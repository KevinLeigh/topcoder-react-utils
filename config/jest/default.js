module.exports = {
  collectCoverage: true,
  collectCoverageFrom: [
    'src/**/*.{js,jsx}',
  ],
  coverageDirectory: '__coverage__',
  moduleNameMapper: {
    '\\.(scss|css)$': 'identity-obj-proxy',
  },
  rootDir: '../..',
  testPathIgnorePatterns: [
    '/node_modules/',
  ],
  transformIgnorePatterns: [
    '/node_modules/(?!topcoder-react-utils)',
  ],
  setupTestFrameworkScriptFile: '<rootDir>/config/jest/setup.js',
};
