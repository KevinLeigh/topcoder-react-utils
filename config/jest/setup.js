/* NOTE: Don't rely on Babel in this module to avoid problems with usage in
 * a host package. */

require('raf/polyfill');
