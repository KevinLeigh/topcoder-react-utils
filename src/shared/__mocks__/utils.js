import * as mock from '../utils';
import isomorphy from '../utils/__mocks__/isomorphy';

mock.isomorphy = isomorphy;

export default mock;
