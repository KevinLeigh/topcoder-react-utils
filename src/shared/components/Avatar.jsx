/**
 * The standard Avatar component.
 * To use it you will have to properly wrap it with react-css-super-themr,
 * the component provided here takes care about the code interface and logic
 * only.
 */

import PT from 'prop-types';
import React from 'react';

function Avatar({ DefaultAvatar, theme, url }) {
  return url
    ? <img alt="Avatar" src={url} className={theme.avatar} />
    : <DefaultAvatar className={theme.avatar} />;
}

Avatar.defaultProps = {
  url: null,
};

Avatar.propTypes = {
  DefaultAvatar: PT.node.isRequired,
  theme: PT.shape({
    avatar: PT.string.isRequired,
  }).isRequired,
  url: PT.string,
};
