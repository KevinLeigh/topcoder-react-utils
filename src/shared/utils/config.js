/**
 * This module provides isomorphic configuration. At the client-side it serves
 * cofiguration object from the global CONFIG variable, injected into the page
 * during the server-side rendering; at the server side it is fetched directly
 * from node-config.
 */

/* eslint-env browser */

import { isClientSide } from './isomorphy';

/* eslint-disable global-require */
module.exports = isClientSide() ? window.CONFIG : require('config');
/* eslint-enable global-require */
