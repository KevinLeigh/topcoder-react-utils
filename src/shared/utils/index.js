import * as isomorphy from './isomorphy';

export default {
  isomorphy,
};
