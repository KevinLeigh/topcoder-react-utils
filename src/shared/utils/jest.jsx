/**
 * Utilities for Jest tests.
 */

/* NOTE: At the moment here is no intention to use this module for runtime
 * execution, only for tests; thus we depend on Jest environment and dev
 * dependencies. */

/* global expect */

import PT from 'prop-types';
import React from 'react';
import TU from 'react-dom/test-utils';

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
/* eslint-enable import/no-extraneous-dependencies */

/**
 * Just an alias for TU.findRenderedDOMComponentWithClass(..).
 * @param {Object} dom
 * @param {String} className
 * @return {Object}
 */
export function findInDomByClass(dom, className) {
  return TU.findRenderedDOMComponentWithClass(dom, className);
}

/**
 * Just an alias for TU.scryRenderedDOMComponentsWithClass(..).
 * @param {Object} dom
 * @param {Stirng} className
 * @return {Array}
 */
export function findInDomManyByClass(dom, className) {
  return TU.scryRenderedDOMComponentsWithClass(dom, className);
}

/**
 * Auxiliary wrapper around ReactJS Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the rendered tree.
 */
export function render(component) {
  return Renderer.create(component).toJSON();
}

/* The Wrapper is necessary for the "renderDom(..)" function, because
 * the "renderIntoDocument(..)" function from "react-dom/test-utils" works
 * only with state components, so we have to wrap our ReactJS components into
 * such Wrapper. */
class Wrapper extends React.Component {
  componentDidMount() {}

  render() {
    return this.props.children;
  }
}

Wrapper.propTypes = {
  children: PT.node.isRequired,
};

/**
 * Renders given component into DOM, using "react-dom/test-utils".
 * @param {Object} component ReactJS component to render.
 * @return {Object} Rendered DOM.
 */
export function renderDom(component) {
  return TU.renderIntoDocument(<Wrapper>{component}</Wrapper>);
}

/**
 * Auxiliary wrapper around ReactJS Shallow Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the shallow component's render tree.
 */
export function shallowRender(component) {
  const renderer = new ShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

/**
 * Renders provided ReactJS component with ReactJS Shallow Test Renderer,
 * creates/tests its snapshot, and returns the render result.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of shallow render.
 */
export function shallowSnapshot(component) {
  const res = shallowRender(component);
  expect(res).toMatchSnapshot();
  return res;
}

/**
 * Renders provided ReactJS component with ReactJS Test Renderer,
 * creates/tests the snapshot, and returns the render.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON render of the component.
 */
export function snapshot(component) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export const simulate = TU.Simulate;
