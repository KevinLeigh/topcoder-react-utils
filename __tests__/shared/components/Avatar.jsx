import Avatar from 'components/Avatar';
import React from 'react';
import { snapshot } from 'utils/jest';

const testTheme = {
  avatar: 'avatarClassName',
};

function DefaultAvatar() {
  return 'DEFAULT_AVATAR';
}

test('Matches snapshots', () => {
  snapshot((
    <Avatar
      DefaultAvatar={DefaultAvatar}
      theme={testTheme}
    />
  ));
  snapshot((
    <Avatar
      DefaultAvatar={DefaultAvatar}
      theme={testTheme}
      url="url/to/the/avatar"
    />
  ));
});
