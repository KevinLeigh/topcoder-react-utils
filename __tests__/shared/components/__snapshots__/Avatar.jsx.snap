// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshots 1`] = `"DEFAULT_AVATAR"`;

exports[`Matches snapshots 2`] = `
<img
  alt="Avatar"
  className="avatarClassName"
  src="url/to/the/avatar"
/>
`;
